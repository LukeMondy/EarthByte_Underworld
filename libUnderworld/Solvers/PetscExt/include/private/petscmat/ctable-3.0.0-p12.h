/* Contributed by - Mark Adams */

#if !defined(__CTABLE300_H)
#define __CTABLE300_H

struct _n_PetscTable {
  PetscInt *keytable;
  PetscInt *table;
  PetscInt count;
  PetscInt tablesize;
  PetscInt head;
};

#include "petscctable.h"
#endif
