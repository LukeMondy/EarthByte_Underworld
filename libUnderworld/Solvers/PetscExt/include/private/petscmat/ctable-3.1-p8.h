/* Contributed by - Mark Adams */

#if !defined(__CTABLE31_H)
#define __CTABLE31_H

struct _n_PetscTable {
  PetscInt *keytable;
  PetscInt *table;
  PetscInt count;
  PetscInt tablesize;
  PetscInt head;
};

#include "petscctable.h"
#endif
