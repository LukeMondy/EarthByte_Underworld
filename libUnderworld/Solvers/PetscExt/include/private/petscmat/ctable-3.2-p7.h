/* Contributed by - Mark Adams */

#if !defined(__CTABLE_H)
#define __CTABLE_H

struct _n_PetscTable {
  PetscInt *keytable;
  PetscInt *table;
  PetscInt count;
  PetscInt tablesize;
  PetscInt head;
};

#include <petscctable.h>
#endif
