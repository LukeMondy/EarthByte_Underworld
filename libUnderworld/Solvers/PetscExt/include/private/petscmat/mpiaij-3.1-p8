
#if !defined(__MPIAIJ_H)
#define __MPIAIJ_H

#include "../src/mat/impls/aij/seq/aij.h"
#include "../src/sys/ctable.h"

typedef struct {
  Mat           A,B;                   /* local submatrices: A (diag part),
                                           B (off-diag part) */
  PetscMPIInt   size;                   /* size of communicator */
  PetscMPIInt   rank;                   /* rank of proc in communicator */ 

  /* The following variables are used for matrix assembly */

  PetscTruth    donotstash;             /* PETSC_TRUE if off processor entries dropped */
  MPI_Request   *send_waits;            /* array of send requests */
  MPI_Request   *recv_waits;            /* array of receive requests */
  PetscInt      nsends,nrecvs;         /* numbers of sends and receives */
  PetscScalar   *svalues,*rvalues;     /* sending and receiving data */
  PetscInt      rmax;                   /* maximum message length */
#if defined (PETSC_USE_CTABLE)
  PetscTable    colmap;
#else
  PetscInt      *colmap;                /* local col number of off-diag col */
#endif
  PetscInt      *garray;                /* global index of all off-processor columns */

  /* The following variables are used for matrix-vector products */

  Vec           lvec;              /* local vector */
  Vec           diag;
  VecScatter    Mvctx;             /* scatter context for vector */
  PetscTruth    roworiented;       /* if true, row-oriented input, default true */

  /* The following variables are for MatGetRow() */

  PetscInt      *rowindices;       /* column indices for row */
  PetscScalar   *rowvalues;        /* nonzero values in row */
  PetscTruth    getrowactive;      /* indicates MatGetRow(), not restored */

  /* Used by MatDistribute_MPIAIJ() to allow reuse of previous matrix allocation  and nonzero pattern */
  PetscInt      *ld;               /* number of entries per row left of diagona block */
} Mat_MPIAIJ;

typedef struct { /* used by MatMatMult_MPIAIJ_MPIAIJ and MatPtAP_MPIAIJ_MPIAIJ for reusing symbolic mat product */
  PetscInt       *startsj,*startsj_r;
  PetscScalar    *bufa;
  IS             isrowa,isrowb,iscolb; 
  Mat            *aseq,*bseq,C_seq; /* A_seq=aseq[0], B_seq=bseq[0] */
  Mat            A_loc,B_seq;
  Mat            B_loc,B_oth;  /* partial B_seq -- intend to replace B_seq */
  PetscInt       brstart; /* starting owned rows of B in matrix bseq[0]; brend = brstart+B->m */
  PetscInt       *abi,*abj; /* symbolic i and j arrays of the local product A_loc*B_seq */
  PetscInt       abnz_max;  /* max(abi[i+1] - abi[i]), max num of nnz in a row of A_loc*B_seq */
  MatReuse       reuse; 
  PetscErrorCode (*MatDestroy)(Mat);
  PetscErrorCode (*MatDuplicate)(Mat,MatDuplicateOption,Mat*);
} Mat_MatMatMultMPI;

typedef struct { /* used by MatMerge_SeqsToMPI for reusing the merged matrix */
  PetscLayout    rowmap;
  PetscInt       **buf_ri,**buf_rj;
  PetscMPIInt    *len_s,*len_r,*id_r; /* array of length of comm->size, store send/recv matrix values */
  PetscMPIInt    nsend,nrecv;  
  PetscInt       *bi,*bj; /* i and j array of the local portion of mpi C=P^T*A*P */
  PetscInt       *owners_co,*coi,*coj; /* i and j array of (p->B)^T*A*P - used in the communication */
  PetscErrorCode (*MatDestroy)(Mat);
  PetscErrorCode (*MatDuplicate)(Mat,MatDuplicateOption,Mat*);
} Mat_Merge_SeqsToMPI; 

EXTERN PetscErrorCode MatSetColoring_MPIAIJ(Mat,ISColoring);
EXTERN PetscErrorCode MatSetValuesAdic_MPIAIJ(Mat,void*);
EXTERN PetscErrorCode MatSetValuesAdifor_MPIAIJ(Mat,PetscInt,void*);
EXTERN PetscErrorCode MatSetUpMultiply_MPIAIJ(Mat);
EXTERN PetscErrorCode DisAssemble_MPIAIJ(Mat);
EXTERN PetscErrorCode MatDuplicate_MPIAIJ(Mat,MatDuplicateOption,Mat *);
EXTERN PetscErrorCode MatIncreaseOverlap_MPIAIJ(Mat,PetscInt,IS [],PetscInt);
EXTERN PetscErrorCode MatFDColoringCreate_MPIAIJ(Mat,ISColoring,MatFDColoring);
EXTERN PetscErrorCode MatGetSubMatrices_MPIAIJ (Mat,PetscInt,const IS[],const IS[],MatReuse,Mat *[]);
EXTERN PetscErrorCode MatGetSubMatrix_MPIAIJ_All(Mat,MatGetSubMatrixOption,MatReuse,Mat *[]);

EXTERN PetscErrorCode MatGetSubMatrix_MPIAIJ(Mat,IS,IS,MatReuse,Mat *);
EXTERN PetscErrorCode MatGetSubMatrix_MPIAIJ_Private (Mat,IS,IS,PetscInt,MatReuse,Mat *);
EXTERN PetscErrorCode MatLoad_MPIAIJ(PetscViewer, const MatType,Mat*);
EXTERN PetscErrorCode MatMatMult_MPIAIJ_MPIAIJ(Mat,Mat,MatReuse,PetscReal,Mat*);
EXTERN PetscErrorCode MatMatMultSymbolic_MPIAIJ_MPIAIJ(Mat,Mat,PetscReal,Mat*);
EXTERN PetscErrorCode MatMatMultNumeric_MPIAIJ_MPIAIJ(Mat,Mat,Mat);
EXTERN PetscErrorCode MatPtAPSymbolic_MPIAIJ(Mat,Mat,PetscReal,Mat*);
EXTERN PetscErrorCode MatPtAPNumeric_MPIAIJ(Mat,Mat,Mat);
EXTERN PetscErrorCode MatPtAPSymbolic_MPIAIJ_MPIAIJ(Mat,Mat,PetscReal,Mat*);
EXTERN PetscErrorCode MatPtAPNumeric_MPIAIJ_MPIAIJ(Mat,Mat,Mat);
EXTERN PetscErrorCode MatSetValues_MPIAIJ(Mat,PetscInt,const PetscInt[],PetscInt,const PetscInt[],const PetscScalar [],InsertMode);
EXTERN PetscErrorCode MatDestroy_MPIAIJ_MatMatMult(Mat);
EXTERN PetscErrorCode PetscContainerDestroy_Mat_MatMatMultMPI(void*);
EXTERN PetscErrorCode MatGetRedundantMatrix_MPIAIJ(Mat,PetscInt,MPI_Comm,PetscInt,MatReuse,Mat*);
EXTERN PetscErrorCode MatGetSeqNonzeroStructure_MPIAIJ(Mat,Mat*);

EXTERN_C_BEGIN
EXTERN PetscErrorCode MatMPIAIJSetPreallocation_MPIAIJ(Mat,PetscInt,const PetscInt[],PetscInt,const PetscInt[]);
EXTERN_C_END

#if !defined(PETSC_USE_COMPLEX) && !defined(PETSC_USE_SCALAR_SINGLE) && !defined(PETSC_USE_SCALAR_MAT_SINGLE)
EXTERN PetscErrorCode MatLUFactorSymbolic_MPIAIJ_TFS(Mat,IS,IS,const MatFactorInfo*,Mat*);
#endif
EXTERN PetscErrorCode MatSolve_MPIAIJ(Mat,Vec,Vec);
EXTERN PetscErrorCode MatILUFactor_MPIAIJ(Mat,IS,IS,const MatFactorInfo *);

EXTERN_C_BEGIN
EXTERN PetscErrorCode PETSCMAT_DLLEXPORT MatGetDiagonalBlock_MPIAIJ(Mat,PetscTruth *,MatReuse,Mat *);
EXTERN PetscErrorCode PETSCMAT_DLLEXPORT MatDiagonalScaleLocal_MPIAIJ(Mat,Vec);

EXTERN_C_END

#endif
