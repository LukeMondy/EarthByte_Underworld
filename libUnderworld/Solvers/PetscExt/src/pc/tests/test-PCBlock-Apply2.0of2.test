EXEC 		= ./test-PCBlock-Apply2.app -sw_pc_type block -pc_block_type upper
NPROCS 		= 2
TEST_TYPE       = tolerance_diff_pl,1.0e-10
