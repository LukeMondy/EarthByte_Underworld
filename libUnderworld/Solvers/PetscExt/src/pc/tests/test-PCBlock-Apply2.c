/*@ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 **
 **    Copyright (c) 2007 - 2008, 
 **        Dave A. May [dave.mayhem23@gmail.com]
 **        School of Mathematical Sciences, 
 **        Monash University
 **        Victoria, 3800
 **        Australia
 **
 **    Copyright (c) 2009 - 2010, 
 **        Dave A. May [dave.mayhem23@gmail.com]
 **        Geophysical Fluid Dynamics, 
 **        Department of Earth Sciences,
 **        ETH ZÃ¼rich,
 **        Switzerland
 **
 **    Project:       PetscExt_v3.1
 **    Description:   A collection of extensions to the PETSc library.
 **
 **    All rights reserved.
 **    Redistribution and use in source and binary forms, with or without modification,
 **    are permitted provided that the following conditions are met:
 **
 **        * Redistributions of source code must retain the above copyright notice,
 **              this list of conditions and the following disclaimer.
 **        * Redistributions in binary form must reproduce the above copyright
 **              notice, this list of conditions and the following disclaimer in the
 **              documentation and/or other materials provided with the distribution.
 **        * Neither the name of the Monash University nor the names of its contributors
 **              may be used to endorse or promote products derived from this software
 **              without specific prior written permission.
 **
 **    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 **    "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
 **    THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 **    PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS
 **    BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 **    CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 **    SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 **    HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 **    LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT 
 **    OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 **
 **
 **    $Id$
 **
 ** ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~@*/


#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <assert.h>

#include <petsc.h>
#include <petscksp.h>
#include <petscmat.h>
#include <petscvec.h>


#include "petscext_vec.h"
#include "petscext_mat.h"
#include "petscext_pc.h"


/*

Q = 
( 1/10   0      0   ) | (  0   3  )
(  0   1/100    0   ) | (  1   4  )
(  0     0   1/1000 ) | (  2   5  )
-----------------------------------
(   0    0      0   ) | ( 1/8  0  ) 
(   0    0      0   ) | (  0  1/9 )

X = ( 1 2 3 | 4 5 )^T
Y = Q^{-1} X 
Y = 
       -1340
      -21000
     -286000
          32
          45
*/
void test_pc_block_apply( void )
{
	int nu,np,i,j;
	Mat A11, A12, A22, A;
	Vec x1,x2, X,Y, y1,y2;
	PC Q11, Q22, Q;
	KSP ksp_Q;	
	
	
	PetscPrintf( PETSC_COMM_WORLD, "%s \n", __func__ );
	nu = 3;
	np = 2;
	
	
	/* Create + Setup A11 */
	MatCreate( PETSC_COMM_WORLD, &A11 );
	MatSetSizes( A11, PETSC_DECIDE, PETSC_DECIDE, nu, nu );
	MatSetType( A11, MATAIJ );
	MatSeqAIJSetPreallocation( A11, nu, PETSC_NULL );
	
	MatSetValue( A11, 0,0, (double)(1.0/10.0), INSERT_VALUES );
	MatSetValue( A11, 1,1, (double)(1.0/100.0), INSERT_VALUES );
	MatSetValue( A11, 2,2, (double)(1.0/1000.0), INSERT_VALUES );
	
	MatAssemblyBegin( A11, MAT_FINAL_ASSEMBLY);
	MatAssemblyEnd( A11, MAT_FINAL_ASSEMBLY);
	/*
	PetscPrintf( PETSC_COMM_WORLD, "A11 = \n");
	PetscViewerSetFormat( PETSC_VIEWER_STDOUT_WORLD, PETSC_VIEWER_ASCII_MATLAB_DENSE );
	MatView( A11, PETSC_VIEWER_STDOUT_WORLD );
	*/
	
	
	/* Create + Setup A12 */
	MatCreate( PETSC_COMM_WORLD, &A12 );
	MatSetSizes( A12, PETSC_DECIDE, PETSC_DECIDE, nu, np );
	MatSetType( A12, MATAIJ );
	MatSeqAIJSetPreallocation( A12, np, PETSC_NULL );
	
	for( j=0; j<np; j++ ) {
		for( i=0; i<nu; i++ ) {
			MatSetValue( A12, i,j, (double)(i+j*nu), INSERT_VALUES );
		}
	}
	MatAssemblyBegin( A12, MAT_FINAL_ASSEMBLY);
	MatAssemblyEnd( A12, MAT_FINAL_ASSEMBLY);
	/*
	PetscPrintf( PETSC_COMM_WORLD, "A12 = \n");
	PetscViewerSetFormat( PETSC_VIEWER_STDOUT_WORLD, PETSC_VIEWER_ASCII_MATLAB_DENSE );
	MatView( A12, PETSC_VIEWER_STDOUT_WORLD );
	*/
	
	
	/* create a matrix for the (2,2) block */
	MatCreate( PETSC_COMM_WORLD, &A22 );
	MatSetSizes( A22, PETSC_DECIDE, PETSC_DECIDE, np, np );
	MatSetType( A22, MATAIJ );
	MatSeqAIJSetPreallocation( A22, np, PETSC_NULL );
	
	MatSetValue( A22, 0,0, (double)(1.0/8.0), INSERT_VALUES );
	MatSetValue( A22, 1,1, (double)(1.0/9.0), INSERT_VALUES );
	
	MatAssemblyBegin( A22, MAT_FINAL_ASSEMBLY);
	MatAssemblyEnd( A22, MAT_FINAL_ASSEMBLY);
	/*
	PetscPrintf( PETSC_COMM_WORLD, "A22 = \n");
	PetscViewerSetFormat( PETSC_VIEWER_STDOUT_WORLD, PETSC_VIEWER_ASCII_MATLAB_DENSE );
	MatView( A22, PETSC_VIEWER_STDOUT_WORLD );
	*/
	
	/* Create + Setup A */
	MatCreate( PETSC_COMM_WORLD, &A );
	MatSetSizes_Block( A, 2,2, 2,2 );
	MatSetType( A, "block" );
	
	MatBlockSetValue( A, 0,0, A11, DIFFERENT_NONZERO_PATTERN, INSERT_VALUES );
	MatBlockSetValue( A, 0,1, A12, DIFFERENT_NONZERO_PATTERN, INSERT_VALUES );
	MatBlockSetValue( A, 1,1, A22, DIFFERENT_NONZERO_PATTERN, INSERT_VALUES );
	MatAssemblyBegin( A, MAT_FINAL_ASSEMBLY);
	MatAssemblyEnd( A, MAT_FINAL_ASSEMBLY);
	
	
	
	/* Create + Setup vectors */
	MatGetVecs( A12, &x2, &x1 );
	MatGetVecs( A12, &y2, &y1 );
	
	/* Setup vectors X */
	VecZeroEntries( x1 );
	VecZeroEntries( x2 );
	for( i=0; i<nu; i++ ) {
		VecSetValue( x1, i, (double)(1+i), INSERT_VALUES );
	}
	VecAssemblyBegin( x1 );
	VecAssemblyEnd( x1 );
	for( i=0; i<np; i++ ) {
		VecSetValue( x2, i, (double)(1+i+nu), INSERT_VALUES );
	}
	VecAssemblyBegin( x2 );
	VecAssemblyEnd( x2 );

	
	
	/* Create block vectors */
	VecCreate( PETSC_COMM_WORLD, &X );
	VecSetSizes( X, 2,2 );
	VecSetType( X, "block" );
	
	VecBlockSetValue( X, 0, x1, INSERT_VALUES );	Stg_VecDestroy( & x1 );
	VecBlockSetValue( X, 1, x2, INSERT_VALUES );	Stg_VecDestroy( & x2 );
	
	VecAssemblyBegin( X );
	VecAssemblyEnd( X );
	
	VecCreate( PETSC_COMM_WORLD, &Y );
	VecSetSizes( Y, 2,2 );
	VecSetType( Y, "block" );
	
	VecBlockSetValue( Y, 0, y1, INSERT_VALUES );	Stg_VecDestroy( & y1 );
	VecBlockSetValue( Y, 1, y2, INSERT_VALUES );	Stg_VecDestroy( & y2 );
	
	VecAssemblyBegin( Y );
	VecAssemblyEnd( Y );
	
	
	
	
	
	/* Block pc */
	PCCreate( PETSC_COMM_WORLD, &Q );
	PCSetOperators( Q, A, A, SAME_NONZERO_PATTERN );
	PCSetOptionsPrefix( Q, "sw_" );
	PCSetFromOptions( Q );
	

#if 0
	PCSetType( Q, "block" );
	PCBlockSetBlockType( Q, PC_BLOCK_UPPER );
	
	/* configure */
	/* Q11 */
	PCBlockGetSubKSP( Q, 0, &ksp_Q );
	KSPGetPC( ksp_Q, &Q11 );
	PCSetType( Q11, "lu" );
	
	/* Q22 */
	PCBlockGetSubKSP( Q, 1, &ksp_Q );
	KSPGetPC( ksp_Q, &Q22 );
	PCSetType( Q22, "lu" );
#endif	
	
	/* Apply preconditioner */
	PCApply( Q, X, Y );
	PCView( Q, PETSC_VIEWER_STDOUT_WORLD );
	
	
//	VecView( Y, PETSC_VIEWER_STDOUT_WORLD );
	VecBlockGetSubVector( Y, 0, &y1 );
	VecBlockGetSubVector( Y, 1, &y2 );
	VecView( y1, PETSC_VIEWER_STDOUT_WORLD );
	VecView( y2, PETSC_VIEWER_STDOUT_WORLD );
	
	
	
	Stg_VecDestroy( & X );	Stg_VecDestroy( & Y );
	Stg_MatDestroy( A11 );	Stg_MatDestroy( A12 );	Stg_MatDestroy( A22 );
	Stg_MatDestroy( A );
	Stg_PCDestroy( Q );
	
}



int main( int argc, char **args )
{
	int i;
	int BIG = 1;
	
	PetscInitialize( &argc, &args, (char *)0, PETSC_NULL );
	PetscExtVecRegisterAll();
	PetscExtMatRegisterAll();
	PetscExtPCRegisterAll();
	
	
	for( i=0; i<BIG; i++ ) {
		test_pc_block_apply();
		
	}
	
	
	PetscExtVecRegisterDestroyAll();
	PetscExtMatRegisterDestroyAll();
	PetscExtPCRegisterDestroyAll();
	
	PetscFinalize();
	return 0;
	
}
