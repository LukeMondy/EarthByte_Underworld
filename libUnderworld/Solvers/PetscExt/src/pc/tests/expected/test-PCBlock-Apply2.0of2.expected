test_pc_block_apply 
PC Object:(sw_)
  type: block
    Block size:   2 x 2 
    BlockPC type: upper 
    PCSetUp_PCBlockApplyTranspose not yet called. 
    PCBlock structure: 
    [TYPE] (i,j) - (Name / ptr), type=, rows=, cols=
      [KSP/PC]  (0,0) - (sw_pc_block_Q11_), type=preonly/bjacobi 
      [Mat] (0,1) - (0x101835810), type=mpiaij, rows=3, cols=2 
      [Mat] (1,0) - (PETSC_NULL) 
      [KSP/PC]  (1,1) - (sw_pc_block_Q22_), type=preonly/bjacobi 
    PCBlock contents: 
    (i,j) - [sub TYPE]: (Name / ptr),  [parent pc]: (Name / ptr) 
        -------------------- Begin ( 0x101824010 ) --------------------
        (0,0) - [sub pc]: (sw_pc_block_Q11_), [parent pc]: sw_ 
        KSP Object:(sw_pc_block_Q11_)
          type: preonly
          maximum iterations=10000, initial guess is zero
          tolerances:  relative=1e-05, absolute=1e-50, divergence=10000
          left preconditioning
          using PRECONDITIONED norm type for convergence test
        PC Object:(sw_pc_block_Q11_)
          type: bjacobi
            block Jacobi: number of blocks = 2
            Local solve is same for all blocks, in the following KSP and PC objects:
          KSP Object:(sw_pc_block_Q11_sub_)
            type: preonly
            maximum iterations=10000, initial guess is zero
            tolerances:  relative=1e-05, absolute=1e-50, divergence=10000
            left preconditioning
            using PRECONDITIONED norm type for convergence test
          PC Object:(sw_pc_block_Q11_sub_)
            type: ilu
              ILU: out-of-place factorization
              0 levels of fill
              tolerance for zero pivot 1e-12
              using diagonal shift to prevent zero pivot
              matrix ordering: natural
              factor fill ratio given 1, needed 1
                Factored matrix follows:
                  Matrix Object:
                    type=seqaij, rows=2, cols=2
                    package used to perform factorization: petsc
                    total: nonzeros=2, allocated nonzeros=10
                      not using I-node routines
            linear system matrix = precond matrix:
            Matrix Object:
              type=seqaij, rows=2, cols=2
              total: nonzeros=2, allocated nonzeros=10
                not using I-node routines
          linear system matrix = precond matrix:
          Matrix Object:
            type=mpiaij, rows=3, cols=3
            total: nonzeros=3, allocated nonzeros=21
              not using I-node (on process 0) routines
        --------------------   End ( 0x101824010 ) --------------------

        -------------------- Begin ( 0x101835810 ) --------------------
        (0,1) - [sub mat]: (0x101835810), [parent pc]: sw_ 
        Matrix Object:
          type=mpiaij, rows=3, cols=2
          total: nonzeros=6, allocated nonzeros=21
            using I-node (on process 0) routines: found 1 nodes, limit used is 5
        --------------------   End ( 0x101835810 ) --------------------

        (1,0) - [sub mat]: (PETSC_NULL), [parent pc]: sw_ 
        
        -------------------- Begin ( 0x101824610 ) --------------------
        (1,1) - [sub pc]: (sw_pc_block_Q22_), [parent pc]: sw_ 
        KSP Object:(sw_pc_block_Q22_)
          type: preonly
          maximum iterations=10000, initial guess is zero
          tolerances:  relative=1e-05, absolute=1e-50, divergence=10000
          left preconditioning
          using PRECONDITIONED norm type for convergence test
        PC Object:(sw_pc_block_Q22_)
          type: bjacobi
            block Jacobi: number of blocks = 2
            Local solve is same for all blocks, in the following KSP and PC objects:
          KSP Object:(sw_pc_block_Q22_sub_)
            type: preonly
            maximum iterations=10000, initial guess is zero
            tolerances:  relative=1e-05, absolute=1e-50, divergence=10000
            left preconditioning
            using PRECONDITIONED norm type for convergence test
          PC Object:(sw_pc_block_Q22_sub_)
            type: ilu
              ILU: out-of-place factorization
              0 levels of fill
              tolerance for zero pivot 1e-12
              using diagonal shift to prevent zero pivot
              matrix ordering: natural
              factor fill ratio given 1, needed 1
                Factored matrix follows:
                  Matrix Object:
                    type=seqaij, rows=1, cols=1
                    package used to perform factorization: petsc
                    total: nonzeros=1, allocated nonzeros=5
                      not using I-node routines
            linear system matrix = precond matrix:
            Matrix Object:
              type=seqaij, rows=1, cols=1
              total: nonzeros=1, allocated nonzeros=5
                not using I-node routines
          linear system matrix = precond matrix:
          Matrix Object:
            type=mpiaij, rows=2, cols=2
            total: nonzeros=2, allocated nonzeros=14
              not using I-node (on process 0) routines
        --------------------   End ( 0x101824610 ) --------------------

  linear system matrix = precond matrix:
  Matrix Object:
    type=block, rows=2, cols=2
        MatBlock structure: 
        (i,j) - (Name / ptr), type=, rows=, cols=
          (0,0) - (0x10183e410), type=block, rows=2, cols=2 
            (0,0) - (0x10183aa10), type=mpiaij, rows=3, cols=3 
            (0,1) - (0x101835810), type=mpiaij, rows=3, cols=2 
            (1,0) - (PETSC_NULL) 
            (1,1) - (0x101831410), type=mpiaij, rows=2, cols=2 
        Block structure: DEFAULT_BLOCK 
        MatBlock contents: 
        (i,j) - [sub mat]: (Name / ptr) , [parent mat]: (Name / ptr) 
            -------------------- Begin ( 0x10183aa10 ) --------------------
            (0,0) - [sub mat]: (0x10183aa10), [parent mat]: 0x10183e410 
            Matrix Object:
              type=mpiaij, rows=3, cols=3
              total: nonzeros=3, allocated nonzeros=21
                not using I-node (on process 0) routines
            --------------------   End ( 0x10183aa10 ) --------------------

            -------------------- Begin ( 0x101835810 ) --------------------
            (0,1) - [sub mat]: (0x101835810), [parent mat]: 0x10183e410 
            Matrix Object:
              type=mpiaij, rows=3, cols=2
              total: nonzeros=6, allocated nonzeros=21
                using I-node (on process 0) routines: found 1 nodes, limit used is 5
            --------------------   End ( 0x101835810 ) --------------------

            (1,0) - [sub mat]: (PETSC_NULL), [parent mat]: 0x10183e410 

            -------------------- Begin ( 0x101831410 ) --------------------
            (1,1) - [sub mat]: (0x101831410), [parent mat]: 0x10183e410 
            Matrix Object:
              type=mpiaij, rows=2, cols=2
              total: nonzeros=2, allocated nonzeros=14
                not using I-node (on process 0) routines
            --------------------   End ( 0x101831410 ) --------------------

-1340
-21000
-286000
32
45
